cypress-io/cypress
# feat(launchpad): Settings Page, Inputs, and Text highlighting

This is a Part 1 for the Settings Page. A followup PR will come in for the rest of the project settings and the user settings. Currently, the new Settings Page components are only accessible via the CT Runner. I've attached screenshots.

Closes the following Sub-tasks for the [Settings Epic](https://cypress-io.atlassian.net/browse/UNIFY-130)
1. [#256](https://cypress-io.atlassian.net/browse/UNIFY-256)
2. [#257](https://cypress-io.atlassian.net/browse/UNIFY-257)
3. [#258](https://cypress-io.atlassian.net/browse/UNIFY-258)

<img width="1337" alt="Screen Shot 2021-08-02 at 2 54 36 AM" src="https://user-images.githubusercontent.com/2801156/127817836-287b7f43-2bd4-4ee1-af5c-04f10fb9f894.png">


### Inputs and Syntax Highlighted Inputs
These features required a lot of CSS refactoring and support for the Inputs and InlineCodeEditor so that they could have Icon Buttons on either side of them. This was particularly challenging because both the syntax highlighted input field and the actual input field need to have the same styling as well as the same Icon positioning. After trying quite a few CSS-only approaches and using PrismJS directly, it was easiest to use the pre-built Prism Editor and just style it to be a disabled one-liner.

**InlineCodeEditor vs Input**
<img width="1311" alt="Screen Shot 2021-08-02 at 3 01 44 AM" src="https://user-images.githubusercontent.com/2801156/127817914-5759c112-6563-4754-8594-a42a1aedc017.png">

**Various Inputs**
<img width="1318" alt="Screen Shot 2021-08-02 at 3 01 37 AM" src="https://user-images.githubusercontent.com/2801156/127828450-1dcf7815-69ff-4c96-866e-d1abea10086d.png">

### Icons
I explored using [vite-purge-icons](https://github.com/antfu/vite-plugin-icons) but decided against using it because it requires the iconify runtime and causes a FOUC when the icon is added and removed to the DOM. This is kind of a bummer because you have to import the SVGs as components. At some point we'll likely try to improve this API, but for now this is what we should use.

Recommended API:
```vue
<template>
  <IconCoffee />
  
  <Input :prefixIcon="IconCoffee"/>
</template>

<script setup>
// Icons are namespaced within a "virtual" import, used to tree-shake the icon library
// The convention is 'virtual:vite-icons/{library}/{icon-name}'
import IconCoffee from 'virtual:vite-icons/mdi/coffee' 
</setup>
```

<img width="1350" alt="Screen Shot 2021-08-02 at 3 04 14 AM" src="https://user-images.githubusercontent.com/2801156/127818109-9825a1a7-19f0-47c0-a2a6-447f733d00aa.png">

### Project Settings
There are a few sections in the project settings. This PR covers three of them:
1. Project ID
2. Record Key
3. Experiments

### Tests
1. We have support for `@testing-library/cypress`
2. All new components have meaningful tests

### Types improvements
This PR also demonstrates the newest type-defined props + emit syntax. You can read about it in detail in the [Type-only defineProps/defineEmit RFC](https://github.com/vuejs/rfcs/blob/master/active-rfcs/0040-script-setup.md#type-only-propsemit-declarations).

Here's a quick example...

#### Type-defined Props + Emit
```ts
defineProps<{
  status: 'success' | 'error' | 'warning'
  message?: string
}>()

defineEmit<{
  click: (e: MouseEvent) => any
}>()
```

#### Default values
This is a great pattern for a lot of components, but default values cannot be defined by an interface. For this, there is the [`withDefaults` method](https://github.com/vuejs/rfcs/blob/master/active-rfcs/0040-script-setup.md#default-props-values-when-using-type-declaration)

```ts
interface Props {
  msg?: string
}

const props = withDefaults(defineProps<Props>(), {
  msg: 'hello'
})
```

## How to test
This PR is best tested by:
1. Using Cypress CT to validate the components `yarn cypress:open` within `@packages/launchpad`
2. Using Cypress Electron to ensure there are no regressions in the UI `yarn dev` within `@packages/launchpad`

## Next Steps
1. We have additional stories to complete in the [Settings Epic](https://cypress-io.atlassian.net/browse/UNIFY-130)
2. The GQL PRs worked on by Lachlan and Tim need to be merged into this PR so that we can hook up the real data into each of the components.
3. We need to expose the Settings page on the app-level (waiting for gql)
